src/main/java/com/bugsnag/BugsnagAppender.java

```diff
@@ -0,0 +1,658 @@
+        List<String> keysToRemove = new ArrayList<String>();
+
+        // Loop through all the keys and put them in the correct tabs
+        for (String key : event.getMDCPropertyMap().keySet()) {
+            if (key.startsWith(LOGGING_CONTEXT_REPORT_PREFIX)) {
+                populateKey(key,
+                        event.getMDCPropertyMap().get(key),
+                        LOGGING_CONTEXT_REPORT_PREFIX,
+                        report);
+                keysToRemove.add(key);
+            } else if (key.startsWith(LOGGING_CONTEXT_THREAD_PREFIX)) {
+                populateKey(key,
+                        event.getMDCPropertyMap().get(key),
+                        LOGGING_CONTEXT_THREAD_PREFIX,
+                        report);
+            }
+        }
+
+        // Remove the report keys so that they won't be associated with any other log message
+        for (String key : keysToRemove) {
+            event.getMDCPropertyMap().remove(key);
+        }
+    }
+
+    /**
+     * Adds the given key/value to the report
+     *
+     * @param key    The key to add
+     * @param value  The value to add
+     * @param prefix The prefix of the key
+     * @param report The report to add the value to
+     */
+    private void populateKey(String key, String value, String prefix, Report report) {
+        if (key.contains(LOGGING_CONTEXT_TAB_SEPARATOR)) {
+            String[] parts = key
+                    .substring(prefix.length())
+                    .split(LOGGING_CONTEXT_TAB_SEPARATOR);
+
+            report.addToTab(parts[0], parts[1], getObjectValue(value));
+        } else {
+            report.addToTab("Context Data", key, getObjectValue(value));
+        }
+    }
+
+    /**
+     * Serializes the contents of the value if required
+     *
+     * @param value The value to serialize
+     * @return The value as a string
+     */
+    private static String getStringValue(Object value) {
+        if (value instanceof  String) {
+            return (String)value;
+        } else {
+            try {
+                return getObjectMapper().writeValueAsString(value);
+            } catch (JsonProcessingException exception) {
+                return value.toString();
+            }
+        }
+    }
+
+    /**
+     * Deserializes the contents of the value
+     *
+     * @param value The value to deserialize
+     * @return The value as an object
+     */
+    private static Object getObjectValue(String value) {
+        try {
+            return getObjectMapper().readValue(value, Map.class);
+        } catch (IOException exception) {
+            // Just return the raw string if it could not be read
+            return value;
+        }
+    }
+
+    /**
+     * @return An object mapper to serialize into the logging context with
+     */
+    protected static ObjectMapper getObjectMapper() {
+        if (mapper == null) {
+            mapper = new ObjectMapper();
+        }
```

Comment:
I think there's a chance that multiple instances could be created here, as this will be called from multiple threads

Reply:
Changed to initialize at the top of the file